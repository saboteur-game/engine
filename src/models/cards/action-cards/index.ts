import ActionCard from "./action-card";
import {
  BoardActionCard,
  MapActionCard,
  RockfallActionCard,
} from "./board-action-cards";
import {
  ToolActionCard,
  BreakActionCard,
  RepairActionCard,
} from "./tool-action-cards";

export {
  ActionCard,
  BoardActionCard,
  MapActionCard,
  RockfallActionCard,
  ToolActionCard,
  BreakActionCard,
  RepairActionCard,
};
