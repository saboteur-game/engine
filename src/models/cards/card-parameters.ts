import { Tools } from "../../constants";
import Position from "../position";
import Player from "../player";

export interface IBoardCardParameters {
  position: Position;
}

export interface IToolActionCardParameters {
  player: Player;
  appliedTo: Tools;
}

type CardParameters = IToolActionCardParameters | IBoardCardParameters;

export default CardParameters;
