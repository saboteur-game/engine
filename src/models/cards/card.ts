import { generateId, Pojo } from "../../utils";
import CardParameters from "./card-parameters";

export enum Status {
  unused = "unused",
  played = "played",
  discarded = "discarded",
}

class Card {
  id: string;
  status: Status;

  constructor() {
    this.id = generateId();
    this.status = Status.unused;
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  play(parameters: CardParameters): Card {
    return this;
  }

  setDiscarded(): void {
    // TODO: Check if the state was unused and throw if it was anything else
    this.status = Status.discarded;
  }

  setPlayed(): void {
    // TODO: Check if the state was unused or played and throw if it was anything else
    this.status = Status.played;
  }

  toJS(): Pojo {
    return {
      id: this.id,
      status: this.status,
    };
  }
}

export default Card;
