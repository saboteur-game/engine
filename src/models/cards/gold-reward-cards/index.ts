import { generateId, Pojo } from "../../../utils";

class GoldRewardCard {
  id: string;
  value: number;

  constructor(value: 1 | 2 | 3) {
    this.id = generateId();
    this.value = value;
  }

  toJS(): Pojo {
    return {
      id: this.id,
      value: this.value,
    };
  }
}

export default GoldRewardCard;
