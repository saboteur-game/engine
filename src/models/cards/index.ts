import { shuffle } from "../../utils";
import { Tools } from "../../constants";
import Card, { Sides } from "./card";
import {
  StartPathCard,
  GoldFinishPathCard,
  RockFinishPathCard,
  PassageCard,
  DeadendCard,
  PathCard,
} from "./path-cards";
import {
  MapActionCard,
  RockfallActionCard,
  BreakActionCard,
  RepairActionCard,
  ActionCard,
} from "./action-cards";

const multiply = (length: number, getValue: () => Card) =>
  new Array(length).fill(undefined).map(() => getValue());
const randomBoolean = () => Math.random() < 0.5;

/*
 * Saboteur cards
 * - 44 path cards
 *  -- 1 starting card
 *  -- 3 finishing cards
 *    --- 1 gold [Connected, 1-2-3-4]
 *    --- 2 rock
 *       ---- 1 [Connected 1-2]
 *       ---- 1 [Connected 1-4]
 *  -- 40 tunnel cards
 *    --- 31 passage cards
 *       ---- 3 passage [Connected 2-4]
 *       ---- 4 passage [Connected 1-3]
 *       ---- 4 passage [Connected 1-4]
 *       ---- 5 passage [Connected 1-2]
 *       ---- 5 passage [Connected 1-2-4]
 *       ---- 5 passage [Connected 1-2-3]
 *       ---- 5 passage [Connected 1-2-3-4]
 *    --- 9 dead-end cards
 *       ---- 1 dead-end [Connected 1]
 *       ---- 1 dead-end [Connected 2]
 *       ---- 1 dead-end [Connected 1-2]
 *       ---- 1 dead-end [Connected 1-3]
 *       ---- 1 dead-end [Connected 1-4]
 *       ---- 1 dead-end [Connected 2-4]
 *       ---- 1 dead-end [Connected 1-2-3]
 *       ---- 1 dead-end [Connected 1-2-4]
 *       ---- 1 dead-end [Connected 1-2-3-4]
 * - 27 action cards
 *  -- 6 map cards
 *  -- 3 rock-fall cards
 *  -- 9 break tool cards
 *    --- 3 pick
 *    --- 3 lamp
 *    --- 3 wagon
 *  -- 9 repair tool cards
 *    --- 2 pick
 *    --- 2 lamp
 *    --- 2 wagon
 *    --- 1 pick and lamp
 *    --- 1 pick and wagon
 *    --- 1 lamp and wagon
 *
 * Notes: Card connections
 *
 *        (1)
 *     +-------+
 *     |       |
 * (4) |       | (2)
 *     |       |
 *     +-------+
 *        (3)
 */

export interface IPlacedCards {
  start: StartPathCard;
  gold: GoldFinishPathCard;
  rock1: RockFinishPathCard;
  rock2: RockFinishPathCard;
}

export const getPlacedCards = (): IPlacedCards => {
  const start = new StartPathCard();
  const gold = new GoldFinishPathCard();
  const rock1 = new RockFinishPathCard([Sides.top, Sides.right]);
  const rock2 = new RockFinishPathCard([Sides.top, Sides.left]);

  [start, gold, rock1, rock2].forEach((card) => card.setPlayed());

  return { start, gold, rock1, rock2 };
};

export const getShuffledDeck = (): Array<ActionCard | PathCard> => {
  const pathCards = [
    ...multiply(
      3,
      () => new PassageCard([Sides.right, Sides.left], randomBoolean())
    ),
    ...multiply(
      4,
      () => new PassageCard([Sides.top, Sides.bottom], randomBoolean())
    ),
    ...multiply(
      4,
      () => new PassageCard([Sides.top, Sides.left], randomBoolean())
    ),
    ...multiply(
      5,
      () => new PassageCard([Sides.top, Sides.right], randomBoolean())
    ),
    ...multiply(
      5,
      () =>
        new PassageCard([Sides.top, Sides.right, Sides.left], randomBoolean())
    ),
    ...multiply(
      5,
      () =>
        new PassageCard([Sides.top, Sides.right, Sides.bottom], randomBoolean())
    ),
    ...multiply(
      5,
      () =>
        new PassageCard(
          [Sides.top, Sides.right, Sides.bottom, Sides.left],
          randomBoolean()
        )
    ),
    new DeadendCard([Sides.top], randomBoolean()),
    new DeadendCard([Sides.right], randomBoolean()),
    new DeadendCard([Sides.top, Sides.right], randomBoolean()),
    new DeadendCard([Sides.top, Sides.bottom], randomBoolean()),
    new DeadendCard([Sides.top, Sides.left], randomBoolean()),
    new DeadendCard([Sides.right, Sides.left], randomBoolean()),
    new DeadendCard([Sides.top, Sides.right, Sides.bottom], randomBoolean()),
    new DeadendCard([Sides.top, Sides.right, Sides.left], randomBoolean()),
    new DeadendCard(
      [Sides.top, Sides.right, Sides.bottom, Sides.left],
      randomBoolean()
    ),
  ] as Array<PathCard>;

  const actionCards = [
    ...multiply(6, () => new MapActionCard()),
    ...multiply(3, () => new RockfallActionCard()),
    ...multiply(3, () => new BreakActionCard(Tools.pick)),
    ...multiply(3, () => new BreakActionCard(Tools.lamp)),
    ...multiply(3, () => new BreakActionCard(Tools.wagon)),
    ...multiply(2, () => new RepairActionCard(Tools.pick)),
    ...multiply(2, () => new RepairActionCard(Tools.lamp)),
    ...multiply(2, () => new RepairActionCard(Tools.wagon)),
    new RepairActionCard([Tools.pick, Tools.lamp]),
    new RepairActionCard([Tools.pick, Tools.wagon]),
    new RepairActionCard([Tools.lamp, Tools.wagon]),
  ] as Array<ActionCard>;

  return shuffle(
    ([] as Array<ActionCard | PathCard>).concat(pathCards, actionCards)
  );
};
