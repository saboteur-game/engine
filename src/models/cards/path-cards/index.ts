import PathCard from "./path-card";
import { TunnelCard, PassageCard, DeadendCard } from "./tunnel-cards";
import {
  StartPathCard,
  FinishPathCard,
  GoldFinishPathCard,
  RockFinishPathCard,
} from "./boundary-cards";

export {
  PathCard,
  TunnelCard,
  PassageCard,
  DeadendCard,
  StartPathCard,
  FinishPathCard,
  GoldFinishPathCard,
  RockFinishPathCard,
};
