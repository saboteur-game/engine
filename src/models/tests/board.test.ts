import Board, { startPosition, middleFinishPosition } from "../board";
import Position from "../position";
import {
  PathCard,
  FinishPathCard,
  StartPathCard,
  PassageCard,
  DeadendCard,
} from "../cards/path-cards";
import { Sides } from "../cards/card";

jest.spyOn(global.Math, "random").mockReturnValue(0.5);

const START_POSITION = startPosition;
const FINISH_POSITION = middleFinishPosition;
const PLAY_POSITION = new Position(0, 1);

const playPassage = (board: Board): void => {
  const passageCard = new PassageCard([
    Sides.top,
    Sides.right,
    Sides.bottom,
    Sides.left,
  ]);
  board.addCard(passageCard, PLAY_POSITION);
};

describe("Board", () => {
  let board: Board;

  beforeEach(() => {
    board = new Board();
  });

  it("can be serialized", () => {
    expect(board.toJS()).toMatchSnapshot();
  });

  describe("visualize", () => {
    beforeEach(() => {
      // Passage up
      board.addCard(
        new PassageCard([Sides.top, Sides.bottom]),
        new Position(0, 1)
      );
      board.addCard(
        new PassageCard([Sides.top, Sides.bottom]),
        new Position(0, 2)
      );
      board.addCard(
        new PassageCard([Sides.top, Sides.bottom]),
        new Position(0, 3)
      );
      board.addCard(
        new PassageCard([Sides.top, Sides.bottom]),
        new Position(0, 4)
      );
      board.addCard(new DeadendCard([Sides.bottom]), new Position(0, 5));

      // Passage right
      board.addCard(
        new PassageCard([Sides.right, Sides.left]),
        new Position(1, 0)
      );
      board.addCard(
        new PassageCard([Sides.right, Sides.left]),
        new Position(2, 0)
      );
      board.addCard(
        new PassageCard([Sides.right, Sides.left]),
        new Position(3, 0)
      );
      board.addCard(
        new PassageCard([Sides.right, Sides.left]),
        new Position(4, 0)
      );
      board.addCard(
        new PassageCard([Sides.top, Sides.left]),
        new Position(5, 0)
      );
      board.addCard(
        new PassageCard([Sides.top, Sides.bottom]),
        new Position(5, 1)
      );
      board.addCard(
        new PassageCard([Sides.right, Sides.bottom]),
        new Position(5, 2)
      );
      board.addCard(
        new PassageCard([Sides.right, Sides.left]),
        new Position(6, 2)
      );

      // Passage down
      board.addCard(
        new PassageCard([Sides.top, Sides.bottom]),
        new Position(0, -1)
      );
      board.addCard(
        new PassageCard([Sides.top, Sides.bottom]),
        new Position(0, -2)
      );
      board.addCard(
        new PassageCard([Sides.top, Sides.bottom]),
        new Position(0, -3)
      );
      board.addCard(
        new PassageCard([Sides.top, Sides.bottom]),
        new Position(0, -4)
      );
      board.addCard(new DeadendCard([Sides.top]), new Position(0, -5));

      // Passage left
      board.addCard(
        new PassageCard([Sides.right, Sides.left]),
        new Position(-1, 0)
      );
      board.addCard(
        new PassageCard([Sides.right, Sides.left]),
        new Position(-2, 0)
      );
      board.addCard(
        new PassageCard([Sides.right, Sides.left]),
        new Position(-3, 0)
      );
      board.addCard(
        new PassageCard([Sides.right, Sides.left]),
        new Position(-4, 0)
      );
      board.addCard(new DeadendCard([Sides.right]), new Position(-5, 0));
    });

    it("returns displayable version of the board", () => {
      expect(board.visualize()).toMatchSnapshot();
    });
  });

  describe("get card from board", () => {
    describe("when start position provided", () => {
      it("returns start card", () => {
        expect(board.getCardAt(START_POSITION) instanceof StartPathCard).toBe(
          true
        );
      });
    });

    describe("when finish position provided", () => {
      it("returns finish card", () => {
        expect(board.getCardAt(FINISH_POSITION) instanceof FinishPathCard).toBe(
          true
        );
      });
    });

    describe("when empty position provided", () => {
      it("returns undefined", () => {
        expect(board.getCardAt(PLAY_POSITION)).toBe(undefined);
      });
    });
  });

  describe("add card to the board", () => {
    describe("when card cannot be added", () => {
      it("throws exception", () => {
        const pathCard = new PathCard([
          Sides.top,
          Sides.right,
          Sides.bottom,
          Sides.left,
        ]);
        expect(() => board.addCard(pathCard, PLAY_POSITION)).toThrow(
          "Invalid type of card provided"
        );
      });
    });

    describe("when card can be added", () => {
      let passageCard: PassageCard;

      beforeEach(() => {
        passageCard = new PassageCard([
          Sides.top,
          Sides.right,
          Sides.bottom,
          Sides.left,
        ]);
      });

      describe("and position is taken", () => {
        beforeEach(() => {
          playPassage(board);
        });

        it("throws exception", () => {
          expect(() => board.addCard(passageCard, PLAY_POSITION)).toThrow(
            "Position 0,1 is already occupied"
          );
        });
      });

      describe("and card is already discarded", () => {
        beforeEach(() => {
          passageCard.setDiscarded();
        });

        it("throws exception", () => {
          expect(() => board.addCard(passageCard, PLAY_POSITION)).toThrow(
            "This card has already been played or discarded"
          );
        });
      });

      describe("and card is already played", () => {
        beforeEach(() => {
          board.addCard(passageCard, new Position(-1, 0));
        });

        it("throws exception", () => {
          expect(() => board.addCard(passageCard, PLAY_POSITION)).toThrow(
            "This card has already been played or discarded"
          );
        });
      });

      describe("and position is empty", () => {
        it("adds card", () => {
          expect(() => board.addCard(passageCard, PLAY_POSITION)).not.toThrow();
          expect(board.getCardAt(PLAY_POSITION)).toBe(passageCard);
        });
      });
    });
  });

  describe("remove card from the board", () => {
    describe("when card cannot be removed from position", () => {
      it("throws exception", () => {
        expect(() => board.removeCard(START_POSITION)).toThrow(
          "Cannot remove card at 0,0"
        );
      });
    });

    describe("when card can be removed from position", () => {
      describe("and position is empty", () => {
        it("throws exception", () => {
          expect(() => board.removeCard(PLAY_POSITION)).toThrow(
            "Position 0,1 is already empty"
          );
        });
      });

      describe("and position is occupied", () => {
        beforeEach(() => {
          playPassage(board);
        });

        it("removes card", () => {
          expect(() => board.removeCard(PLAY_POSITION)).not.toThrow();
          expect(board.getCardAt(PLAY_POSITION)).toBe(undefined);
        });
      });
    });
  });

  describe("get adjacent cards from the board", () => {
    let passageCard: PassageCard;

    describe("when the position is beside the start card", () => {
      beforeEach(() => {
        passageCard = new PassageCard([Sides.right, Sides.left]);
        board.addCard(passageCard, new Position(2, 0));
      });

      it("returns the start card and other surrounding cards", () => {
        expect(board.getAdjacentCards(new Position(1, 0))).toMatchSnapshot();
      });
    });

    describe("when the position is beside a finish card", () => {
      beforeEach(() => {
        passageCard = new PassageCard([Sides.right, Sides.left]);
        board.addCard(passageCard, new Position(5, 0));
      });

      it("returns surrounding cards but not the finish card", () => {
        expect(board.getAdjacentCards(new Position(6, 0))).toMatchSnapshot();
      });
    });

    describe("when the position is beside nothing", () => {
      beforeEach(() => {
        passageCard = new PassageCard([Sides.right, Sides.left]);
        board.addCard(passageCard, new Position(1, 0));
      });

      it("returns all empty spaces", () => {
        expect(board.getAdjacentCards(new Position(3, 0))).toMatchSnapshot();
      });
    });

    describe("when the position is surrounded", () => {
      const getDeadendCrossroads = () =>
        new DeadendCard([Sides.top, Sides.right, Sides.bottom, Sides.left]);

      const getPassageCrossroads = () =>
        new PassageCard([Sides.top, Sides.right, Sides.bottom, Sides.left]);

      beforeEach(() => {
        board.addCard(getDeadendCrossroads(), new Position(0, 1));
        board.addCard(getPassageCrossroads(), new Position(1, 1));
        board.addCard(getDeadendCrossroads(), new Position(2, 1));
        // 0,0 is starting point and 1,0 is the position querying adjacent to
        board.addCard(getPassageCrossroads(), new Position(2, 0));
        board.addCard(getDeadendCrossroads(), new Position(0, -1));
        board.addCard(getPassageCrossroads(), new Position(1, -1));
        board.addCard(getDeadendCrossroads(), new Position(2, -1));
      });

      it("returns surrounding cards", () => {
        expect(board.getAdjacentCards(new Position(1, 0))).toMatchSnapshot();
      });
    });
  });
});
