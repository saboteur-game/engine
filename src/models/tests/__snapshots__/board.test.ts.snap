// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board can be serialized 1`] = `
Object {
  "grid": Object {
    "0,0": StartPathCard {
      "connectors": Array [
        1,
        2,
        3,
        4,
      ],
      "id": "test-id-6",
      "isUpsideDown": false,
      "status": "played",
    },
    "7,-2": RockFinishPathCard {
      "connectors": Array [
        1,
        4,
      ],
      "id": "test-id-9",
      "isUpsideDown": false,
      "status": "played",
    },
    "7,0": RockFinishPathCard {
      "connectors": Array [
        1,
        2,
      ],
      "id": "test-id-8",
      "isUpsideDown": false,
      "status": "played",
    },
    "7,2": GoldFinishPathCard {
      "connectors": Array [
        1,
        2,
        3,
        4,
      ],
      "id": "test-id-7",
      "isUpsideDown": false,
      "status": "played",
    },
  },
  "id": "test-id-5",
}
`;

exports[`Board get adjacent cards from the board when the position is beside a finish card returns surrounding cards but not the finish card 1`] = `
Array [
  undefined,
  undefined,
  undefined,
  PassageCard {
    "connectors": Array [
      2,
      4,
    ],
    "id": "test-id-284",
    "isUpsideDown": false,
    "status": "played",
  },
]
`;

exports[`Board get adjacent cards from the board when the position is beside nothing returns all empty spaces 1`] = `
Array [
  undefined,
  undefined,
  undefined,
  undefined,
]
`;

exports[`Board get adjacent cards from the board when the position is beside the start card returns the start card and other surrounding cards 1`] = `
Array [
  undefined,
  PassageCard {
    "connectors": Array [
      2,
      4,
    ],
    "id": "test-id-272",
    "isUpsideDown": false,
    "status": "played",
  },
  undefined,
  StartPathCard {
    "connectors": Array [
      1,
      2,
      3,
      4,
    ],
    "id": "test-id-268",
    "isUpsideDown": false,
    "status": "played",
  },
]
`;

exports[`Board get adjacent cards from the board when the position is surrounded returns surrounding cards 1`] = `
Array [
  PassageCard {
    "connectors": Array [
      1,
      2,
      3,
      4,
    ],
    "id": "test-id-310",
    "isUpsideDown": false,
    "status": "played",
  },
  PassageCard {
    "connectors": Array [
      1,
      2,
      3,
      4,
    ],
    "id": "test-id-314",
    "isUpsideDown": false,
    "status": "played",
  },
  PassageCard {
    "connectors": Array [
      1,
      2,
      3,
      4,
    ],
    "id": "test-id-318",
    "isUpsideDown": false,
    "status": "played",
  },
  StartPathCard {
    "connectors": Array [
      1,
      2,
      3,
      4,
    ],
    "id": "test-id-304",
    "isUpsideDown": false,
    "status": "played",
  },
]
`;

exports[`Board visualize returns displayable version of the board 1`] = `
"                              ██████                                          
                              ██████                                          
                              ██  ██                                          
                              ██  ██                                          
                              ██  ██                                          
                              ██  ██                                          
                              ██  ██                                          
                              ██  ██                                          
                              ██  ██                                          
                              ██  ██                        ██████████████  ██
                              ██  ██                        ██            !!  
                              ██  ██                        ██  ██████████  ██
                              ██  ██                        ██  ██            
                              ██  ██                        ██  ██            
                              ██  ██                        ██  ██            
████████████████████████████████  ████████████████████████████  ██      ██  ██
████                            []                              ██      ██xx  
████████████████████████████████  ████████████████████████████████      ██████
                              ██  ██                                          
                              ██  ██                                          
                              ██  ██                                          
                              ██  ██                                    ██  ██
                              ██  ██                                      xx██
                              ██  ██                                    ██████
                              ██  ██                                          
                              ██  ██                                          
                              ██  ██                                          
                              ██  ██                                          
                              ██  ██                                          
                              ██  ██                                          
                              ██  ██                                          
                              ██████                                          
                              ██████                                          "
`;
